icaruscode/Generators/HepMCFileGen_module.cc

```diff
@@ -0,0 +1,457 @@
+ *  The use of `TLorentzVector` below does not imply space and time have the same units
+ *   (do not use `TLorentzVector::Boost()`).
+ */
+#include <string>
+#include <fstream>
+#include <math.h>
+#include <map>
+#include <iomanip>
+#include <algorithm>
+#include <sstream>
+#include <glob.h>
+#include <cstdlib>  // for unsetenv()
+#include "art/Framework/Core/EDProducer.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "cetlib_except/exception.h"
+#include "art/Framework/Services/Registry/ServiceHandle.h"
+#include "cetlib/search_path.h"
+#include "cetlib/getenv.h"
+#include "cetlib/split_path.h"
+#include "cetlib_except/exception.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+#include "TLorentzVector.h"
+#include "larcore/Geometry/Geometry.h"
+#include "larcoreobj/SummaryData/RunData.h"
+#include "larcoreobj/SummaryData/POTSummary.h"
+#include "nusimdata/SimulationBase/MCTruth.h"
+#include "nusimdata/SimulationBase/MCParticle.h"
+#include "ifdh.h"       // use bare
+#undef USE_IFDH_SERVICE // ifdh for now
+// #ifndef NO_IFDH_LIB
+//   #define USE_IFDH_SERVICE 1
+//   // IFDHC
+//   #ifdef USE_IFDH_SERVICE
+//     #include "IFDH_service.h"
+//   #else
+//     // bare IFDHC
+//     #include "ifdh.h"
+//   #endif
+// #else
+//   #undef USE_IFDH_SERVICE
+//   // nothing doing ... use ifdef to hide any reference that might need header
+//   #include <cassert>
+// #endif
+namespace evgen {
+  class HepMCFileGen;
+}
+class evgen::HepMCFileGen : public art::EDProducer {
+public:
+  explicit HepMCFileGen(fhicl::ParameterSet const & p);
+  void produce(art::Event & e)        override;
+  void beginJob()                                 override;
+  void beginRun(art::Run & run)                   override;
+  void endSubRun(art::SubRun& sr)     override;
+private:
+//  void ExpandInputFilePatternsDirect();
+//  void ExpandInputFilePatternsIFDH();
+  void open_file();
+  std::ifstream* fInputFile;
+//  std::string              fFileSearchPaths; ///< colon separated set of path stems (to be set)
+//  std::vector<std::string> fFilePatterns;    ///< wildcard patterns files containing histograms or ntuples, or txt (to be set)
+//  std::vector<std::string> fSelectedFiles;   ///< flux files selected after wildcard expansion and subset selection
+//  std::string              fFileCopyMethod;  ///< "DIRECT" = old direct access method, otherwise = ifdh approach schema ("" okay)
+  
+  double         fEventsPerPOT;     ///< Number of events per POT (to be set)
+  int            fEventsPerSubRun;  ///< Keeps track of the number of processed events per subrun
+  ifdh_ns::ifdh* fIFDH;             ///< (optional) flux file handling
+};
+//------------------------------------------------------------------------------
+evgen::HepMCFileGen::HepMCFileGen(fhicl::ParameterSet const & p)
+  : EDProducer{p}
+  , fInputFile(0)
```

Comment:
To have the file name as parameter (2):
```suggestion
  , fInputFilePath(p.get<std::string>("InputFilePath"))
  , fInputFile(nullptr)
```
```suggestion
  , fInputFile(0)
```